index.js

```diff
@@ -109,7 +150,22 @@ Plugin.prototype.toString = function() {
 // This indirection allows subclasses like broccoli-caching-writer to hook
 // into calls from the builder, by returning { build: someFunction }
 Plugin.prototype.getCallbackObject = function() {
-  return this
+  var plugin = this;
+
+  if (plugin._sideEffectFree === false) {
+    return this;
+  }
+
+  return {
+    build: function() {
```

Comment:
We should move this method to be on the `Plugin.prototype` because it is totally valid to extend from `Plugin` and implement a custom `getCallbackObject`.  In that case, our guard logic should still be available to be used if desired...